Let incoming webhook define or redefine parameters

chmouel opened the pull request:
Let users pass parameters to the incoming webhook.

We accept now JSON payloads in the incoming webhook and we can pass the params to the pipelinerun dynamically.

Allowed params to be passed need to be explicitly defined in the Repository CR as an extra security layer.

See docs for more informations

### Demo

https://github.com/openshift-pipelines/pipelines-as-code/assets/98980/61eba60c-90d9-499c-aff7-f6440cfcc745



<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [x] 📝 A good commit message is important for other reviewers to understand the context of your change. Please refer to [How to Write a Git Commit Message](https://cbea.ms/git-commit/) for more details how to write beautiful commit messages. We rather have the commit message in the PR body and the commit message instead of an external website.
- [x] ♽ Run `make test` before submitting a PR (ie: with [pre-commit](https://pipelinesascode.com/dev/tools), no need to waste CPU cycle on CI. (or even better install [pre-commit](https://pre-commit.com/) and do `pre-commit install` in the root of this repo).
- [x] ✨ We heavily rely on linters to get our code clean and consistent, please ensure that you have run `make lint` before submitting a PR. The [markdownlint](https://github.com/DavidAnson/markdownlint) error can get usually fixed by running `make fix-markdownlint` (make sure it's installed first)
- [x] 📖 If you are adding a user facing feature or make a change of the behavior, please verify that you have documented it
- [x] 🧪 100% coverage is not a target but most of the time we would rather have a unit test if you make a code change.
- [x] 🎁 If that's something that is possible to do please ensure to check if we can add a e2e test.
- [x] 🔎 If there is a flakiness in the CI tests then don't *necessary* ignore it, better get the flakyness fixed before merging or if that's not possible there is a good reason to bypass it. (token rate limitation may be a good reason to skip).

chmouel committed:
Let incoming webhook define or redefine parameters

Let users pass parameters to the incoming webhook.

We accept now JSON payloads in the incoming webhook and we can pass the
params to the pipelinerun dynamically.

Allowed params to be passed need to be explicitly defined in the
Repository CR to avoid for an extra security layer.

See docs for more informations

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
weird the e2e works locally but not on ci 🤔 


```
No resources found
Running TestGithubIncomingWebhook
Cleaning test cache
=== RUN   TestGithubIncomingWebhook
💡 13:27:44 Testing GithubApp Incoming with Github APPS integration on pac-e2e-ns-gg7xp
💡 13:27:45 Namespace pac-e2e-ns-gg7xp created
💡 13:27:45 PipelinesAsCode Repository pac-e2e-ns-gg7xp has been created in namespace pac-e2e-ns-gg7xp
💡 13:27:46 Commit e7db8a56231e63341627743bc678cde22dc28f16 has been created and pushed to branch refs/heads/pac-e2e-ns-gg7xp
💡 13:27:50 Kicked off on incoming URL: http://controller.paac-127-0-0-1.nip.io/incoming?repository=pac-e2e-ns-gg7xp&branch=pac-e2e-ns-gg7xp&pipelinerun=pipelinerun-incoming&secret=shhhh-secrete
💡 13:27:50 Waiting for Repository to be updated
💡 13:27:50 Repo number of statuses: 0
💡 13:27:51 Repo number of statuses: 0
💡 13:27:52 Repo number of statuses: 0
💡 13:27:53 Repo number of statuses: 0
💡 13:27:54 Repo number of statuses: 0
💡 13:27:55 Repo number of statuses: 0
💡 13:27:56 Repo number of statuses: 1
💡 13:27:56 Check if we have the repository set as succeeded
💡 13:27:56 looking for regexp .*It's a Bird... It's a Plane... It's Superman in pipelinesascode.tekton.dev/event-type=incoming:step-task labelSelector/pod
💡 13:27:56 matched regexp in labelSelector/container pipelinesascode.tekton.dev/event-type=incoming:step-task
💡 13:27:56 Closing PR -1
💡 13:27:56 Deleting Repository in pac-e2e-ns-gg7xp
💡 13:27:56 Deleting NS pac-e2e-ns-gg7xp
💡 13:27:56 Deleting Ref refs/heads/pac-e2e-ns-gg7xp
--- PASS: TestGithubIncomingWebhook (12.46s)
PASS
ok  	github.com/openshift-pipelines/pipelines-as-code/test	12.494s
```

chmouel commented:
/test go-testing

chmouel commented:
/retest

savitaashture commented on pkg/adapter/incoming.go:
```diff
@@ -88,6 +98,29 @@ func (l *listener) detectIncoming(ctx context.Context, req *http.Request, payloa
+		return false, nil, fmt.Errorf("invalid content type, only application/json is accepted when posting a body")
+	}
+
+	payload, err := apincoming.ParseIncomingPayload(payloadBody)
```
I think we should handle marshaling error here and return if there is an error?

chmouel replied:
the current default documented behaviour  is an empty POST, this is to handle backward compatibility, but maybe it doesn't matter since the json decoder will handle empty string

savitaashture commented on pkg/adapter/incoming_test.go:
```diff
@@ -143,6 +148,47 @@ func Test_listener_detectIncoming(t *testing.T) {
 			},
 		},
+		{
+			name: "good/incoming with body",
```
can we have one more test scenario where multiple params in repo CR but passed json has only one matching param which is again a green scenario

I mean
Repo CR
```
spec:
  incoming:
  - params:
    - test
    - test1111
```

Curl request

```
curl -H "Content-Type: application/json" -X POST "https://url/incoming?repository=savitaashture-article&branch=main&secret=very-secure-shared-secret&pipelinerun=article-no-operation" -d '{"params": {"test1111": "12345"}}'
```

savitaashture replied:
@chmouel is it addressed ?

chmouel replied:
i just opushed it!

savitaashture commented on pkg/apis/incoming/incoming.go:
```diff
@@ -0,0 +1,17 @@
+}
+
+// parsePayload parses the payload from the incoming webhook, in json format and has only one key params
+func ParseIncomingPayload(payload []byte) (Payload, error) {
```
Why this function is pkg/apis :thinking: 

Is it okay if we keep in pkg/customparams/incoming.go ? or someother place 


Also unit test is pending

chmouel replied:
because of cycling dependencies...

chmouel committed:
GH App, make sure token generated works for itself

When we were generating a token we would do this for the installation
token where the repo is installed.  However, some installations would
not have the repo where it comes from (or target on incoming webhook)
and due fail.

Fixed the tests along the way which wasn't tested thing fully and made
them more robust.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

savitaashture commented:
/lgtm
